feat(Transaction Deletion Record): Editable "DocTypes To Delete" List with CSV import/export

HenningWendtland opened the pull request:
## Summary

This PR improves the Transaction Deletion Record feature by adding an editable "DocTypes To Delete" list with CSV import/export capabilities, providing better user control and reusability across environments.

Please consider to Backport this to v15

## Why?
The intention is to create a UI Tool for preparing production environments from test setups or to strip data from a production environment to (re)create a test setup. The requirements of which data to keep vary from case to case and only roughly align with the existence of a "company" field. The current implementation does not give the user any manual control on what will be deleted.
In Multi Company setups it is also common to create extra "company" fields, which makes deletion based on that trigger more complicated, so some control is desired

## Changes

### 1. Editable "DocTypes To Delete" List
- **New Child DocType**: `Transaction Deletion Record To Delete`
  - Fields include: `doctype_name`, `company_field`, `document_count` (read-only), `child_doctypes` (read-only), `deleted` (progress tracking)
  - Supports multiple company fields per DocType
- **User Workflow**: 
  1. Generate To Delete list automatically (filters by company field + exclusions)
  2. Review and manually edit the list (add/remove DocTypes)
  3. Submit to start deletion
- **Validation**: Prevents protected DocTypes and child tables from being added
- **Backwards compatability**: keeps the current "doctypes_to_be_ignored" table and does not add DocTypes to  "DocTypes To Delete" List
- **Protected Doctypes**: Adds a list of protected Frappe Core Doctypes that cannot be deleted 

### 2. CSV Template Import/Export
- **Export**: Download To Delete list as CSV template (includes Child Doctypes for reference)
- **Import**: Upload a list of Doctypes in CSV format to populate list. 
- **Use Case**: Save and reuse deletion templates across environments (dev/test/prod)
- **Validation**: CSV header validation, child table filtering, consolidated skip messages

### 3. UX changes
- **Conditional Visibility**: Summary table only shows for legacy submitted records without To Delete list
- **Auto-Start**: Deletion starts automatically after submit confirmation, submit message warns user about potential data loss

### 4. Code Quality
- Added concise docstrings to all new methods
- Single API entry point per feature (removed redundant `@frappe.whitelist()` decorators)

## Backward Compatibility

- **Existing Records**: Continue to work with original auto-generation logic
- **Summary Table**: Still visible for historical/legacy records
- **Fallback Logic**: Maintained for records without To Delete list

## Technical Details

### New Methods
- `generate_to_delete_list()` - Auto-populate To Delete list
- `populate_doctype_details(doctype_name, company)` - Get details for manually added DocTypes
- `export_to_delete_template(name)` - Module-level CSV export function (whitelisted)
- `import_to_delete_template_method(csv_content)` - Internal CSV import handler
- `process_import_template(transaction_deletion_record_name, file_url)` - Module-level CSV import function (whitelisted)

### Modified Behaviour
- `before_submit()` now requires To Delete list for new workflow
- `initialize_doctypes_to_be_deleted_table()` uses To Delete list when available, falls back to legacy logic
- `delete_company_transactions()` marks "deleted" checkbox in To Delete list when done

## Testing

- [x] Generate To Delete list works correctly
- [x] Manual add/remove DocTypes validated properly
- [x] Export CSV includes correct data
- [x] Import CSV regenerates counts and validates
- [x] Child tables are filtered during import
- [x] Protected DocTypes are rejected
- [x] Backward compatibility with existing records
- [x] Summary table visibility logic works

## Screenshots
<img width="1000" height="552" alt="screenshot_2025-11-18-3-small" src="https://github.com/user-attachments/assets/f8af1134-5bcb-4f57-8223-686669cb97b4" />

<img width="1000" height="549" alt="screenshot_2025-11-18-2-small" src="https://github.com/user-attachments/assets/c8109930-73ce-4fa2-8e38-b40593072848" />

<img width="1000" height="372" alt="screenshot_2025-11-18-1-small" src="https://github.com/user-attachments/assets/19774f0e-22c5-439c-aaeb-e2eedd62963a" />

<img width="1000" height="510" alt="screenshot_2025-11-18-4-small" src="https://github.com/user-attachments/assets/690616e9-9f38-4472-a1f6-4bd8cf9ef946" />


## Documentation

https://docs.frappe.io/erpnext/user/manual/en/delete_company_transactions?editWiki=1&wikiPagePatch=bnj7ckhqfa

HenningWendtland committed:
feat: add editable DocTypes To Delete list with import/export

Add user control over transaction deletion with reviewable and reusable deletion templates.

- New "DocTypes To Delete" table allows users to review and customize what will be deleted before submission
- Import/Export CSV templates for reusability across environments
- Company field rule: only filter by company if field is specifically named "company", otherwise delete all records
- Child tables (istable=1) automatically excluded from selection
- "Remove Zero Counts" helper button to clean up list
- Backward compatible with existing deletion records

HenningWendtland committed:
refactor: improve Transaction Deletion Record code quality

- Remove unnecessary chatty comments from AI-generated code
- Add concise docstrings to all new methods
- Remove redundant @frappe.whitelist() decorators from internal methods
- Improve CSV import validation (header check, child table filtering)
- Add better error feedback with consolidated skip messages
- Reorder form fields: To Delete list now appears before Excluded list
- Add conditional visibility for Summary table (legacy records only)
- Improve architectural clarity: single API entry point per feature

Technical improvements:
- export_to_delete_template_method and import_to_delete_template_method
  are now internal helpers without whitelist decorators
- CSV import now validates format and provides detailed skip reasons
- Summary table only shows for submitted records without To Delete list
- Maintains backward compatibility for existing deletion records

HenningWendtland committed:
fix: field order

HenningWendtland committed:
test: fix broken tests and add new ones

HenningWendtland committed:
fix: adapt create_transaction_deletion_request

HenningWendtland committed:
test: fix assertRaises trigger

HenningWendtland committed:
fix: conditionally execute Transaction Deletion pre-tasks based on selected DocTypes

HenningWendtland committed:
refactor: replace boolean task flags with status fields

HenningWendtland committed:
fix: remove UI comment

HenningWendtland committed:
fix: don't allow virtual doctype selection and improve protected Doctype List

HenningWendtland committed:
fix: replace outdated frappe.db.sql by frappe.qb

HenningWendtland committed:
feat: add support for multiple company fields

HenningWendtland committed:
fix: autofill comapny field, add docstrings, filter for company_field

HenningWendtland committed:
fix: add edge case handling for update_naming_series and add tests for prefix extraction

HenningWendtland commented:
I realised that the function `check_for_running_deletion_job` that is called generally on document validate does not fit together well with this PR:
I want to explicitly allow the System Manager to delete  DocTypes without "company" field, but a general validation would use up performance unnecessarily

Proposed solution:
Remove this validation entirely. The **Transaction Deletion Record** should be used carefully and infrequently and often inside a test environment or even before Going Live with a production system. I already implemented a prominent message asking the user to back up data beforehand.

barredterra commented:
Discussed with @sagarvora:

In a larger setup ("Big Company" scenario), a System Manager might start a deletion job while other employees are still entering data or testing features. Since the deletion task can take significant time (e.g., ~25 mins), there is a large window for race conditions. If we remove the check, we risk data corruption or users complaining that records weren't actually deleted because new data came in during the process.

Proposed Compromise: Instead of removing it, can we make the check cheaper using Redis or [globals](https://github.com/frappe/frappe/blob/f95905cbaac8e01c8c9b0345be60cb5cf14a7efb/frappe/database/database.py#L1138-L1144):

1. On Process Start: Set a key/flag.
2. On Validate: Check for this key (almost zero performance cost).
3. On Completion: Unset the key.

This keeps the system foolproof against concurrent writes without the heavy DB overhead.

sagarvora commented:
option B - no regular checks needed:

after deletion check whether everything was deleted successfully. if not:
- retry full process, OR
- show error

HenningWendtland committed:
fix: use redis for running deletion validation, check per doctype instead of company

HenningWendtland commented:
> Discussed with @sagarvora:
> 
> In a larger setup ("Big Company" scenario), a System Manager might start a deletion job while other employees are still entering data or testing features. Since the deletion task can take significant time (e.g., ~25 mins), there is a large window for race conditions. If we remove the check, we risk data corruption or users complaining that records weren't actually deleted because new data came in during the process.
> 
> Proposed Compromise: Instead of removing it, can we make the check cheaper using Redis or [globals](https://github.com/frappe/frappe/blob/f95905cbaac8e01c8c9b0345be60cb5cf14a7efb/frappe/database/database.py#L1138-L1144):
> 
> 1. On Process Start: Set a key/flag.
> 2. On Validate: Check for this key (almost zero performance cost).
> 3. On Completion: Unset the key.
> 
> This keeps the system foolproof against concurrent writes without the heavy DB overhead.

I implemented redis cache usage for the validation. I removed the filtering by company but instead introduced one flag per Doctype that is deleted - such that DocTypes that are not being deleted are not affected by the validation

HenningWendtland commented:
@ruthra-kumar linter failure is unrelated to this PR

ruthra-kumar commented:
@HenningWendtland Can you rebase on latest develop.

ruthra-kumar commented:
@HenningWendtland 
 Nice!

ruthra-kumar commented:
Will backport to v16 on Feb
